openmdao/components/external_code_comp.py

```diff
@@ -9,80 +9,77 @@
 
 from openmdao.core.analysis_error import AnalysisError
 from openmdao.core.explicitcomponent import ExplicitComponent
+from openmdao.core.implicitcomponent import ImplicitComponent
 from openmdao.utils.shell_proc import STDOUT, DEV_NULL, ShellProc
 from openmdao.utils.general_utils import warn_deprecation
 
 
-class ExternalCodeComp(ExplicitComponent):
+class ExternalCodeDelegate(object):
     """
-    Run an external code as a component.
-
-    Default stdin is the 'null' device, default stdout is the console, and
-    default stderr is ``error.out``.
+    Handles all the methods related to running a code externally.
 
     Attributes
     ----------
-    stdin : str or file object
-        Input stream external code reads from.
-    stdout : str or file object
-        Output stream external code writes to.
-    stderr : str or file object
-        Error stream external code writes to.
-    DEV_NULL : File object
-        NULL device.
-    STDOUT : File object
-        Special value that can be used as the stderr argument to Popen and indicates
-        that standard error should go into the same handle as standard output.
-    return_code : int
-        Exit status of the child process.
+    _comp : ExternalCodeComp or ExternalCodeImplicitComp object
+        The external code object this delegate is associated with.
+
     """
 
-    def __init__(self, **kwargs):
+    def __init__(self, comp):
         """
-        Intialize the ExternalCodeComp component.
+        Initialize.
 
         Parameters
         ----------
-        **kwargs : dict of keyword arguments
-            Keyword arguments that will be mapped into the Component options.
+        comp : ExternalCodeComp or ExternalCodeImplicitComp object
+            The external code object this delegate is associated with.
+
         """
-        super(ExternalCodeComp, self).__init__(**kwargs)
+        self._comp = comp
 
-        self.stdin = DEV_NULL
-        self.stdout = None
-        self.stderr = "external_code_comp_error.out"
+    def init_comp(self):
+        """
+        Initialize the external code component this delegate is associated with.
+        """
+        comp = self._comp
 
-        self.DEV_NULL = DEV_NULL
```

Comment:
I know you are trying to save duplicate code here, but I think it might be better to move these attributes declarations back into the ExernalCodeComp and ExternalImplicitCodeComp's __init__.  That seems more pythonic to me, and it won't confuse the linters.

Reply:
do these even need to be public attributes?  Seems like they could just be private attributes of the delegate...  if any of these are intended to be user-facing (they are not mentioned in the doc) then they should probably be an option?

Reply:
Ken, I debated that also. I was surprised the linters didn't complain. I will put them back into the __init__